### DISCUSSION TITLE
Am I being limited by single core CPU performance when fully offloaded?

### DISCUSSION THREAD

**Discussion by reversebias** [ggml-org/llama.cpp#5803] [Q&A]
I'm using a system with the following hardware:

- Xeon W-2133
- 96GB DDR4-2666
- 2x Nvidia P40 each in a 16x 3.0 slot 

Running a Q6 quant of mixtral with:
`./main -m ~/models/mixtral-8x7b-instruct-v0.1.Q6_K.gguf -c 32768 -ngl 128 -ts 39,61,0 -sm row -t 8 --prompt "Once upon a time"`

And getting very decent 21-22 t/s:
`llama_print_timings:      sample time =     155.18 ms /   352 runs   (    0.44 ms per token,  2268.29 tokens per second)
llama_print_timings: prompt eval time =     168.00 ms /     5 tokens (   33.60 ms per token,    29.76 tokens per second)
llama_print_timings:        eval time =   16486.27 ms /   351 runs   (   46.97 ms per token,    21.29 tokens per second)
llama_print_timings:       total time =   16912.39 ms /   356 tokens`

However I noticed that during generation, there's a single CPU thread pegged at 100%:
![image](https://github.com/ggerganov/llama.cpp/assets/5522931/4275e9ef-56da-44a4-ae15-bfc9a82416d5)

Single threaded is expected here, from looking at commits like #5238

But with the single thread at 100%, and nvidia-smi showing 50-60% utilization:
![image](https://github.com/ggerganov/llama.cpp/assets/5522931/be9c7802-93f2-4df4-a1ed-8f37bd68a098)

Could my single thread performance be my bottleneck?

**Comment by reversebias** [comment]
Found a response here that suggests it's spinning a thread and not doing anything, so not the bottleneck: https://github.com/ggerganov/llama.cpp/discussions/3210#discussioncomment-7020324

**Reply by ggerganov** [reply]
The thread spinning referenced in the comment is no longer occurring with latest `ggml-backend` updates

**Comment by ExtReMLapin** [comment]
I feel like i'm having the same issue 

![image](https://github.com/ggerganov/llama.cpp/assets/3909752/230be340-9cad-4e7d-a952-2ef053959d45)

**Comment by Artefact2** [comment]
Similar behaviour here on AMD/rocBLAS.

![image](https://github.com/ggerganov/llama.cpp/assets/90720/3c137ddc-9995-4c63-90ac-d6e1c27efb01)

Most of the CPU time is spent in `rocr::core::InterruptSignal::WaitRelaxed(hsa_signal_condition_t, long, unsigned long, hsa_wait_state_t)`. Needless to say, this makes the CPU generate a lot of heat and consume extra power for nothing.

**Comment by moritzschaefer** [comment]
I would have suspected that more CPUs should at least benefit initial tokenization (is that the `load` part of the timings?).

**Comment by jwhitehorn** [comment]
I too am running into this behavior. Single CPU thread at 100%, and GPU under-utilized (about 20% utilization).

Not sure if it matters, but here are some details:

* Debian 12 / 6.8.4 host
* Dual Xeon E5 2697v2 CPUs
* 64GB ECC RAM (Quad-channel DDR3-1333)
* Intel Arc A770 GPU
* Llama.cpp (via llama-cpp-python 0.2.65) dockerized using the intel/oneapi-basekit:2024.1.0-devel-ubuntu22.04 image
* I've tried different models (llama 2, llama 3, claude 2, etc), all fully offloaded to VRAM

I've tried using llama.cpp's main and server executables directly and can confirm that the fact that I'm running through Python seems immaterial to the issue.

If the CPU thread isn't the bottleneck, do we know why the GPU isn't more utilized?


<img width="1538" alt="Screenshot 2024-04-29 at 7 57 23 PM" src="https://github.com/ggerganov/llama.cpp/assets/523322/b7bccb64-6941-49c4-9556-f1bbf0203e29">
<img width="1538" alt="Screenshot 2024-04-29 at 7 57 28 PM" src="https://github.com/ggerganov/llama.cpp/assets/523322/fcdc7dc7-4c80-43e8-ba9e-1353f52dedee">

**Comment by fidecastro** [comment]
I'm also seeing similar behavior on my system. My situation might make it a bit more interesting a problem since I wasn't expecting to see *both* a single CPU thread and GPU maxing out, as I'm seeing here.

Currently running the latest llama.cpp release (b3484) on the following machine:

2x Xeon E5-2690v4 (total 28 cores, 56 threads)
128 GB DDR4 RAM 2133
4 x NVIDIA P40 (3 on 16x PCIe 3.0, 1 on 8x PCIe 3.0)

I compiled llama.cpp with CUDA support on a Docker container based on [this](https://hub.docker.com/layers/nvidia/cuda/12.5.1-cudnn-devel-ubuntu22.04/images/sha256-6f0ad658b45e0468d4de5b3d464cec34b2173343061a667b10877a584c229b77?context=explore) Docker image (Ubuntu 22.04 with CUDA 12.5.1 installed), and which is running with full access to all host GPUs and CPU cores. I also installed flash-attention and am running q4_K_M llama3.1-70B-Instruct.

I'm calling llama-cli with the following terminal command: ```./llama-cli -m /workspace/llama3.1/Meta-Llama-3.1-70B-Instruct-Q4_K_M.gguf -ctk q4_0 -ctv q4_0 -fa -ngl 999 -e --chat-template llama3 -n 8000 --file prompt.txt```

HTOP tells me there's a single CPU with 100% utilization, while my GPUs cycle between idle (3 of the 4 GPUs are always like that) and 100% use (only 1 GPU is in use at a given time).

**Reply by slaren** [reply]
You may get better results with `-sm row` with the P40.

**Comment by stevenbedrick** [comment]
Just chiming in that I am _also_ seeing the same behavior! Model fully offloaded across multiple GPUs (A40s in my case), and then during generation I'm seeing a single CPU thread, pegged at 100, with the GPUs being  underutilized. I've verified that the behavior persists regardless of split mode. 🤷‍♂️

**Comment by 8XXD8** [comment]
I have the same problem, with 4X AMD GPUs, one thread maxed out, and low GPU utilization.
For me, `--no-mmap` helps.

**Reply by stevenbedrick** [reply]
Hmm, thanks; in my case that doesn't seem to make much of a difference. 🤷‍♂️

**Comment by ClonedPoro** [comment]
Is there any updates on this?/ Have you found a solution or workaround? Im seeing the exact same behaviour with a fully offloaded model (still got dedicated memory to spare on the GPU) and barely any GPU utilization but 100% on CPU core 4 in my case (rest is mostly idle).
During initilization my model reads:
AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | AVX512_BF16 = 0 | FMA = 1 | NEON = 0 | SVE = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | MATMUL_INT8 = 0 | LLAMAFILE = 1 |

**Comment by AGenchev** [comment]
My experimets: Performance on AMD APU 5600G (w BIOS GPU overclock)
I'm using same model in all tests (llama-13-v2/ggml-model-q8_0.gguf).
using Vulkan ~ 1.63 tokens per second (but strange - no load on the CPU). 
using CPU - 2.86 tokens per second (around 599% load on CPU). 
using RoCM HIPBLAS - 3.07 tokens per second (1 CPU core loaded at 100%).

So inference with Vulkan comes 'for free' using only the GPU. But it is slowest. Other frameworks are faster but load the CPU.

Haven't tried what it can do with OpenCL.

**Comment by KutayKoray** [comment]
Hi have you done any particular thing to use 2 gpu at the same time because i have 2 1660 super but i cant use both of them model is only recognize 1 gpu. When i serve ollama it detects both of my gpu's when i run nvidia-smi it detects both of my gpu's but when i run llama3.1 it only use one of them has any one faced with this problem or is this because of my gpu model?

**Comment by gilbrotheraway** [comment]
has always been the normal behavior for me on llamacpp when fully offloaded

**Comment by kstoykov** [comment]
I'm still experience the issue.

**Comment by mknap** [comment]
This behavour has been noted in [Discussion 22238](https://github.com/ggml-org/llama.cpp/discussions/22238) which suggests using`--poll 0` as a possible work-around. There is also a patch provided nvidia/cuda setups. 

I am bumping in here because I am seeing this behavior on SYCL backend with Arc Pro B70. I haven't quite got to the bottom of it yet.

**Reply by ExtReMLapin** [reply]
Move to vllm for better performance

**Comment by kstoykov** [comment]
See the accepted answer here: https://github.com/ggml-org/llama.cpp/discussions/22238
